### PR TITLE
Hook system

### PR DESCRIPTION
**Pull request by LuaFan2**
**Experimental hook system**

This system gives modifications creators opportunity to hook game events without any difficulties. 

**Proof of Concept**

```lua
hook:Add("onLevelSwitch", function(level)
    level:discover()
end)
```

vs a work with actors, which includes work with map

**Conclusion**
Game events catching is easier with hooking system that i present, isn't it?

**W**ork**I**n**Progress**
TODO:

- [ ] Opportunity to suppress game events/modify they behavior using hook returns

- [ ] More hooks

- [ ] Optimization

### CODE REVIEW

File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/RemixedDungeon.java [hunk 1]
```diff
@@ -48,15 +49,19 @@
...
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import com.nyrds.pixeldungeon.modding.Event;
 
 import javax.microedition.khronos.opengles.GL10;
 
 public class RemixedDungeon extends Game {
 
 	public static final double[] MOVE_TIMEOUTS = new double[]{250, 500, 1000, 2000, 5000, 10000, 30000, 60000, Double.POSITIVE_INFINITY };
+	public static List<Event> events = new ArrayList<Event>();
```
Comment (Mikhael-Danilov) [inline RemixedDungeon/src/main/java/com/watabou/pixeldungeon/RemixedDungeon.java:64]: Map<String, List<Event>>?

File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/EquipableItem.java [hunk 2]
```diff
@@ -83,6 +84,7 @@ public float time2equip(@NotNull Char hero ) {
 
 	public boolean doEquip(@NotNull Char hero ) {
 		Belongings belongings = hero.getBelongings();
+		new Hook().Call("onArmorEquip", this, hero);
```
Comment (Mikhael-Danilov) [inline RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/EquipableItem.java:87]: onItemEquip, slot, this, hero

File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/EquipableItem.java [hunk 3]
```diff
@@ -123,7 +125,8 @@ protected boolean doUnequip(Char hero, boolean collect, boolean single) {
 		if (collect && !collect( hero.getBelongings().backpack )) {
 			doDrop(hero);
 		}
-				
+
+		new Hook().Call("onArmorUnequip", this, hero);
```
Comment (Mikhael-Danilov) [inline RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/EquipableItem.java:129]: As in previous

File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/Gold.java [hunk 4]
```diff
@@ -66,6 +67,7 @@ public boolean doPickUp(@NotNull Char hero ) {
 		hero.spendAndNext( TIME_TO_PICK_UP );
 		
 		Sample.INSTANCE.play( Assets.SND_GOLD, 1, 1, Random.Float( 0.9f, 1.1f ) );
+		new Hook().Call("onGoldPickup", this, hero);
```
Comment (Mikhael-Danilov) [inline RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/Gold.java:70]: This is really needed when onItemPickup already here?

File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/food/Food.java [hunk 5]
```diff
@@ -51,6 +52,7 @@
 	public void execute(@NotNull Char chr, @NotNull String action ) {
 		if (action.equals( CommonActions.AC_EAT )) {
 			chr.eat(this, energy, message);
+			new Hook().Call("onFoodEat", this, chr);
```
Comment (Mikhael-Danilov) [inline RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/food/Food.java:55]: Will not work for custom items

File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/modding/Hook.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+        for (Object object : objects) {
+            values.add(CoerceJavaToLua.coerce(object));
+        }
+
+        return values;
+    }
+
+    @LuaInterface
+    public void Call(String event, Object... args) {
```
Comment (Mikhael-Danilov) [inline RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/modding/Hook.java:26]: static ?

File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/modding/Hook.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            if(ev.event.equals(event)) {
+                List<LuaValue> values = objectsToValues(objects);
+
+                ev.execute(values.toArray(new LuaValue[values.size()]));
+            }
+        }
+    }
+
+    @LuaInterface
+    public void Add(String name, LuaClosure callback) {
```
Comment (Mikhael-Danilov) [inline RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/modding/Hook.java:40]: static ?

File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/modding/Hook.java (hunk 6 above)
Comment (LuaFan2) [reply RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/modding/Hook.java:26]: Yes, thanks. In next commit.

File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/modding/Hook.java (hunk 7 above)
Comment (LuaFan2) [reply RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/modding/Hook.java:40]: Yes, thanks. In next commit.

File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/food/Food.java (hunk 5 above)
Comment (LuaFan2) [reply RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/food/Food.java:55]: Ok. Thx.

File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/Gold.java (hunk 4 above)
Comment (LuaFan2) [reply RemixedDungeon/src/main/java/com/watabou/pixeldungeon/items/Gold.java:70]: You are right. I will remove it.

File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/RemixedDungeon.java (hunk 1 above)
Comment (LuaFan2) [reply RemixedDungeon/src/main/java/com/watabou/pixeldungeon/RemixedDungeon.java:64]: Why second parameter is List?May be Map<String, Void> where first param is name of event and second is callback?

Comment (LuaFan2) [reply RemixedDungeon/src/main/java/com/watabou/pixeldungeon/RemixedDungeon.java:64]: Yes, in next commit

**Pull request merged**